qiskit/circuit/controlflow/for_loop.py

```diff
@@ -0,0 +1,104 @@
+# This code is part of Qiskit.
+#
+# (C) Copyright IBM 2021.
+#
+# This code is licensed under the Apache License, Version 2.0. You may
+# obtain a copy of this license in the LICENSE.txt file in the root directory
+# of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+#
+# Any modifications or derivative works of this code must retain this
+# copyright notice, and modified files need to carry a notice indicating
+# that they have been altered from the originals.
+
+"Circuit operation representing a ``for`` loop."
+
+import warnings
+from typing import Iterable, Optional
+
+from qiskit.circuit.parameter import Parameter
+from qiskit.circuit.exceptions import CircuitError
+from qiskit.circuit.quantumcircuit import QuantumCircuit
+from .control_flow import ControlFlowOp
+
+
+class ForLoopOp(ControlFlowOp):
+    """A circuit operation which repeatedly executes a subcircuit
+    (``body``) parameterized by a parameter ``loop_parameter`` through
+    the set of integer values provided in ``indexset``.
+
+    Parameters:
+        loop_parameter: The placeholder parameterizing ``body`` to which
+            the values from ``indexset`` will be assigned.
+        indexset: A collection of integers to loop over.
+        body: The loop body to be repeatedly executed.
+        label: An optional label for identifying the instruction.
+
+    **Circuit symbol:**
+
+    .. parsed-literal::
+
+             ┌───────────┐
+        q_0: ┤0          ├
+             │           │
+        q_1: ┤1          ├
+             │  for_loop │
+        q_2: ┤2          ├
+             │           │
+        c_0: ╡0          ╞
+             └───────────┘
+
+    """
+
+    def __init__(
+        self,
+        loop_parameter: Parameter,
+        indexset: Iterable[int],
+        body: QuantumCircuit,
+        label: Optional[str] = None,
+    ):
+        num_qubits = body.num_qubits
+        num_clbits = body.num_clbits
+
+        super().__init__(
+            "for_loop", num_qubits, num_clbits, [loop_parameter, indexset, body], label=label
+        )
+
+    @property
+    def params(self):
+        return self._params
+
+    @params.setter
+    def params(self, parameters):
+        loop_parameter, indexset, body = parameters
```

Comment:
Is it worth checking if `parameters` can unpack and, if not, giving a nicer error message, like the ones below?

Reply:
Good question, I think https://github.com/Qiskit/qiskit-terra/pull/7123#discussion_r748317092 addresses, but let me know if not!